### PR TITLE
[PLATFORM-861]: Fix borked enum variant name formatting

### PR DESCRIPTION
**Pull request by WilliamVenner**
https://prima-assicurazioni-spa.myjetbrains.com/youtrack/issue/PLATFORM-861

Fixes a weird edge case where enum variant names were not being formatted correctly.

Added tests to ensure that we don't regress this in future.

### CODE REVIEW

File: src/private.rs [hunk 1]
```diff
@@ -1,35 +1,24 @@
...
-
-pub struct RedactFlags {
+pub enum RedactSpecialization {
     /// Whether the type we're redacting is an Option<T> or not. Poor man's specialization! This is detected
     /// by the proc macro reading the path to the type, so it's not perfect.
     ///
     /// This could be improved & rid of in a number of different ways in the future:
     ///
-    /// * Once specialization is stabilized, we can use a trait to override redacting behaviour for some types,
+    /// * Once specialization is stabilized, we can use a trait to override redacting behavior for some types,
```
Comment (MaeIsBad) [inline src/private.rs:9]: I can't believe a british dev would do this. Did it hurt?

File: src/private.rs [hunk 2]
```diff
@@ -52,14 +41,14 @@ impl RedactFlags {
     /// Maximum number of characters to expose at the beginning and end of a partial redact.
     const MAX_PARTIAL_EXPOSE: usize = 3;
 
-    fn redact_partial(&self, str: &str, redacted: &mut String) {
+    pub(crate) fn redact_partial(&self, fmt: &mut std::fmt::Formatter, str: &str) -> std::fmt::Result {
```
Comment (MaeIsBad) [inline src/private.rs:44]: I don't like the non descriptive `str`, maybe a `to_redact` would be better

File: src/private.rs [hunk 3]
```diff
@@ -52,14 +41,14 @@ impl RedactFlags {
...
         let count = str.chars().filter(|char| char.is_alphanumeric()).count();
         if count < Self::MIN_PARTIAL_CHARS {
             for char in str.chars() {
                 if char.is_alphanumeric() {
-                    redacted.push(self.redact_char);
+                    fmt.write_char(self.redact_char)?;
```
Comment (MaeIsBad) [inline src/private.rs:49]: I'm a bit worried about the performance of using many write_chars instead of batching the writes together. No metrics on this though

File: src/private.rs [hunk 4]
```diff
@@ -72,100 +61,118 @@ impl RedactFlags {
...
                 }
             }
         }
+        Ok(())
     }
 
-    fn redact_full(&self, str: &str, redacted: &mut String) {
+    pub(crate) fn redact_full(&self, fmt: &mut std::fmt::Formatter, str: &str) -> std::fmt::Result {
         for char in str.chars() {
             if char.is_whitespace() || !char.is_alphanumeric() {
```
Comment (MaeIsBad) [inline src/private.rs:81]: This is the first time I've seen anyone use `is_alphanumeric` over `is_ascii_alphanumeric` correctly

File: src/private.rs [hunk 5]
```diff
@@ -72,100 +61,118 @@ impl RedactFlags {
...
-                // This should never happen, but just in case...
-                flags.redact_full(&redactable_string, &mut redacted);
+        if self.flags.fixed > 0 {
+            return RedactFlags::redact_fixed(fmt, self.flags.fixed as usize, self.flags.redact_char);
+        }
+
+        let redactable_string = self.this.to_string();
+
+        #[allow(clippy::single_match)]
+        match self.specialization {
```
Comment (MaeIsBad) [inline src/private.rs:150]: why match instead of just if let?

File: veil-macros/src/flags.rs [hunk 6]
```diff
@@ -103,89 +85,87 @@ impl FieldFlags {
...
-                }
-                _ => return Err(syn::Error::new_spanned(attr, "unknown modifier for `Redact` attribute")),
-            }
-        }
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct RedactFlags {
+    /// Whether the field is partially or fully redacted.
+    ///
+    /// Incompatible with `fixed`.
+    pub partial: bool,
```
Comment (MaeIsBad) [inline veil-macros/src/flags.rs:104]: Could we instead make fixed and partial an enum?
Like:
```
enum RedactedSize {
Fixed(NonZeroU8),
Parital,
}
```
?

File: veil-macros/src/lib.rs [hunk 7]
```diff
@@ -1,87 +1,31 @@
+//! Macros for `veil`
+
+#![deny(missing_docs)]
```
Comment (MaeIsBad) [inline veil-macros/src/lib.rs:3]: This could be a warning for development, the CI won't pass if the warnings are present due to `cargo clippy -- -D warnings`

File: veil-macros/src/lib.rs [hunk 8]
```diff
@@ -1,87 +1,31 @@
...
+#![deny(rustdoc::broken_intra_doc_links)]
```
Comment (MaeIsBad) [inline veil-macros/src/lib.rs:4]: This is not checked in CI, we should add
`cargo doc --document-private-items --workspace --all-features --no-deps` 
as a step(copied from `prima_datadog.rs`)

File: src/private.rs (hunk 1 above)
Comment (WilliamVenner) [reply src/private.rs:9]: 💀

File: src/private.rs (hunk 3 above)
Comment (WilliamVenner) [reply src/private.rs:49]: Yeah I was also concerned about that actually. One would hope that there would be a buffered writer in the middle somewhere. Will need to investigate

File: src/private.rs (hunk 4 above)
Comment (WilliamVenner) [reply src/private.rs:81]: Unicode slay

File: src/private.rs (hunk 3 above)
Comment (MaeIsBad) [reply src/private.rs:49]: I guess that depends on the formatter impl. I'm not sure if we should be doing our own buffering, if a formatter doesn't do buffering it's not our problem. On the other hand we could maybe preallocate memory more efficiently? Idk

File: src/private.rs (hunk 5 above)
Comment (WilliamVenner) [reply src/private.rs:150]: We may introduce new specialisations in future, so it would look kind of weird to have an `if let`, which kinda implies there will only ever be one kind of specialisation. Unfortunately `#[non_exhaustive]` has no effect too in the same crate, so that doesn't fix the clippy warning :(

I'll slap on a `#[non_exhaustive]` anyway

Comment (MaeIsBad) [reply src/private.rs:150]: Moved to https://github.com/primait/veil/pull/60 where it belongs

**Comment by MaeIsBad** [comment]
Moved my review over to https://github.com/primait/veil/pull/60/ where it belongs

File: veil-macros/src/enums.rs [hunk 9]
```diff
@@ -168,7 +168,7 @@ pub(super) fn derive_redact(
     for (variant, flags) in e.variants.iter().zip(variant_flags.into_iter()) {
         // Variant name redacting
         let variant_name = variant.ident.to_string();
-        let variant_name = if let Some(flags) = &flags.variant_flags {
+        let variant_name = if let Some(flags @ FieldFlags { skip: false, .. }) = &flags.variant_flags {
```
Comment (cpiemontese) [inline veil-macros/src/enums.rs:171]: This is for me: wth is `@`?

**Review by cpiemontese — APPROVED** [review]

File: veil-macros/src/enums.rs (hunk 9 above)
Comment (MaeIsBad) [reply veil-macros/src/enums.rs:171]: it binds the flags variable if the pattern on the right matches
https://doc.rust-lang.org/stable/rust-by-example/flow_control/match/binding.html

Comment (WilliamVenner) [reply veil-macros/src/enums.rs:171]: such an awesome feature :)
